packages/opentelemetry-exporter-prometheus/src/prometheus.ts

```diff
@@ -0,0 +1,280 @@
+    ).replace(/^([^/])/, '/$1');
+
+    if (config.startServer || PrometheusExporter.DEFAULT_OPTIONS.startServer) {
+      this.startServer(callback);
+    } else if (callback) {
+      callback();
+    }
+  }
+
+  /**
+   * Save current metric state so that it can be pulled by the metrics backend.
+   *
+   * @todo reach into metrics to pull metric values on endpoint
+   * In its current state, the exporter saves the current values of all metrics when export
+   * is called and returns them when the export endpoint is called. In the future, this should
+   * be a no-op and the exporter should reach into the metrics when the export endpoint is
+   * called. As there is currently no interface to do this, this is our only option.
+   *
+   * @param readableMetrics Metrics to be sent to the prometheus backend
+   * @param cb result callback to be called on finish
+   */
+  export(
+    readableMetrics: ReadableMetric[],
+    cb: (result: ExportResult) => void
+  ) {
+    if (!this._server) {
+      // It is conceivable that the _server may not be started as it is an async startup
+      // However unlikely, if this happens the caller may retry the export
+      cb(ExportResult.FAILED_RETRYABLE);
+      return;
+    }
+
+    this._logger.debug('Prometheus exporter export');
+
+    for (const readableMetric of readableMetrics) {
+      this._updateMetric(readableMetric);
+    }
+
+    cb(ExportResult.SUCCESS);
+  }
+
+  /**
+   * Shut down the export server
+   */
+  shutdown() {
+    this.stopServer();
+  }
+
+  /**
+   * Save the value of one metric to be exported
+   *
+   * @param readableMetric Metric value to be saved
+   */
+  private _updateMetric(readableMetric: ReadableMetric) {
+    const metric = this._registerMetric(readableMetric);
+    if (!metric) return;
+
+    const labelKeys = readableMetric.descriptor.labelKeys;
+
+    if (metric instanceof Prometheus.Counter) {
+      for (const ts of readableMetric.timeseries) {
+        // Prometheus counter saves internal state and increments by given value.
+        // ReadableMetric value is the current state, not the delta to be incremented by.
+        // Currently, _registerMetric creates a new counter every time the value changes,
+        // so the increment here behaves as a set value (increment from 0)
+        metric.inc(this._getLabelValues(labelKeys, ts.labelValues), ts.points[0]
+          .value as number);
+      }
+    }
+
+    if (metric instanceof Prometheus.Gauge) {
+      for (const ts of readableMetric.timeseries) {
+        metric.set(this._getLabelValues(labelKeys, ts.labelValues), ts.points[0]
+          .value as number);
+      }
+    }
+
+    // TODO: only counter and gauge are implemented in metrics so far
+  }
+
+  private _getLabelValues(keys: string[], values: LabelValue[]) {
+    const labelValues: Prometheus.labelValues = {};
+    for (let i = 0; i < keys.length; i++) {
+      if (values[i].value !== null) {
```

Comment:
Do label sets allow for unspecified/null values? I know that [unspecified keys](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-metrics-user.md#missing-label-keys) are allowed for aggregation purposes

Reply:
The prometheus labelValues interface is

```typescript
interface labelValues {
    [key: string]: string | number;
}
```

The null check was simply to satisfy this constraint. Our LabelValue interface is

```typescript
interface LabelValue {
    /** The value for the label. */
    readonly value: string | null;
}
```

which you can see does allow null values. Is there a better solution than dropping them? Possibly I should set them to `null`?

WDYT

Reply:
per SIG, dropping values is what open census does so we'll try it for now and see if there are issues